File: marketplace-client-old/src/app/_service/task-interaction.service.ts
import {Injectable} from '@angular/core';
import {Task} from '../_model/task';
import {HttpClient} from '@angular/common/http';
import {Participant} from '../_model/participant';

@Injectable()
export class TaskInteractionService {

  constructor(private http: HttpClient) {
  }

  findByTask(task: Task) {
    return this.http.get('/marketplace/task/' + task.id + '/interaction');
  }

  findFinishedByTask(task: Task) {
    return this.http.get('/marketplace/task/' + task.id + '/interaction?operation=FINISHED');
  }

  findReservedVolunteersByTask(task: Task) {
    return this.http.get(`/marketplace/task/${task.id}/participant?operation=RESERVED`);
  }

  findAssignedVolunteersByTask(task: Task) {
    return this.http.get(`/marketplace/task/${task.id}/participant?operation=ASSIGNED`);
  }

}

File: marketplace/marketplace-application/src/main/java/at/jku/cis/iVolunteer/marketplace/task/interaction/TaskInteractionService.java
package at.jku.cis.iVolunteer.marketplace.task.interaction;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import at.jku.cis.iVolunteer.model.task.Task;
import at.jku.cis.iVolunteer.model.task.interaction.TaskInteraction;
import at.jku.cis.iVolunteer.model.task.interaction.TaskVolunteerOperation;
import at.jku.cis.iVolunteer.model.user.User;
import at.jku.cis.iVolunteer.model.user.Volunteer;

@Service
public class TaskInteractionService {

	@Autowired
	private TaskInteractionRepository taskInteractionRepository;

	public Set<Volunteer> findReservedVolunteersByTask(Task task) {
		Set<Volunteer> volunteers = new HashSet<>();

		findByTaskGroupedByParticipant(task).entrySet().forEach(entry -> {
			List<TaskInteraction> interactions = entry.getValue();
			if (TaskVolunteerOperation.RESERVED == interactions.get(interactions.size() - 1).getOperation()
					|| TaskVolunteerOperation.UNASSIGNED == interactions.get(interactions.size() - 1).getOperation()) {
				volunteers.add((Volunteer) entry.getKey());
			}
		});

		return volunteers;
	}

	public Set<Volunteer> findAssignedVolunteersByTask(Task task) {
		Set<Volunteer> volunteers = new HashSet<>();

		findByTaskGroupedByParticipant(task).entrySet().forEach(entry -> {
			List<TaskInteraction> interactions = entry.getValue();
			if (TaskVolunteerOperation.ASSIGNED == interactions.get(interactions.size() - 1).getOperation()) {
				volunteers.add((Volunteer) entry.getKey());
			}
		});

		return volunteers;
	}

	private Map<User, List<TaskInteraction>> findByTaskGroupedByParticipant(Task task) {
		List<TaskInteraction> taskInteractions = taskInteractionRepository.findByTask(task);
		return taskInteractions.stream().filter(ti -> ti.getParticipant() != null).collect(Collectors.groupingBy(t -> t.getParticipant()));
	}

}
